### ISSUE TITLE
Compilation fails because of incorrect type inference, the analyzer doesn't emit any errors.

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/sdk#48149]
I expected the compiler to be able to infer the type of `invalidExample` and to compile this code without any errors. 

The analyzer can infer the type of `invalidExample` with `add type annotation` to `HBaseComposite<HKindValid, HBase<HKindValid>>` and when added to the code, the compiler stops emitting any errors during compilation.

```dart

const invalidExample = HBaseComposite(
  [
    HBaseLeaf(0),
    HBaseChild(
      HBaseLeaf(0),
    ),
    HBaseError("error message"),
  ],
);

abstract class HBase<HKT extends HKind> implements Kind<HKT> {}

class HBaseComposite<HKT extends HKindValid, CHILD extends HBase<HKT>> implements HBase<HKT> {
  final List<CHILD> children;

  const HBaseComposite(
    final this.children,
  );
}

class HBaseChild<HKT extends HKindValid, CHILD extends HBase<HKT>> implements HBase<HKT> {
  final CHILD child;

  const HBaseChild(
    final this.child,
  );
}

class HBaseLeaf<HKT extends HKindValid> implements HBase<HKT> {
  final int data;

  const HBaseLeaf(
    final this.data,
  );
}

class HBaseError<HKT extends HKindInvalid> implements HBase<HKT> {
  final String errorMessage;

  const HBaseError(
    final this.errorMessage,
  );
}

abstract class Kind<HKT extends HKind> {}

abstract class HKind {}

abstract class HKindInvalid implements HKind {}

abstract class HKindValid implements HKindInvalid {}
```

```
... Error: Type argument 'Object' doesn't conform to the bound 'HBase<HKT>' of the type variable 'CHILD' on 'HBaseComposite'.
 - 'Object' is from 'dart:core'.
 - 'HBase' is from '...' ('...').
Try changing type arguments so that they conform to the bounds.
const invalidExample = HBaseComposite(
      ^
...:13:46: Context: This is the type variable whose bound isn't conformed to.
class HBaseComposite<HKT extends HKindValid, CHILD extends HBase<HKT>> implements HBase<HKT> {
                                             ^
```

**Comment by modulovalue** [comment]
It appears to me that the "add type annotation" feature has been changed to no longer infer the more precise type. 

I'm going to close this issue as there's nothing confusing to me here anymore. It would be nice if type inference could infer the type here, but, admittedly (and sadly), there's not many people that would benefit from that.

The following issue is relevant here as this is an example of where type inference could be improved wrt simulated higher kinded types: https://github.com/dart-lang/language/issues/844

**Comment by eernstg** [comment]
I hadn't looked at this one, but here's a very late comment on the situation:

First, it is working as intended.

The core reason for this behavior is the standard upper bound computation, which is already the topic of a whole [set of issues](https://github.com/dart-lang/language/issues?q=is%3Aopen+is%3Aissue+label%3Aleast-upper-bound) in the language repository. (It's the "standard upper bound", not "least upper bound", because it isn't least. In many cases, a least upper bound does not even exist in Dart, or in any language with multiple subtyping for that matter.)

The difficulties are created by the overly general results produced by the standard upper bound algorithm [`UP`](https://github.com/dart-lang/language/blob/main/resources/type-system/upper-lower-bounds.md#upper-bounds) in some special cases. In particular, no attempt is made to perform recursive invocations of the same algorithm when we're seeking the standard upper bound of two types of the form `A<T1 .. Tk>` and `B<S1 .. Sn>` where `A` and `B` is not the same  generic class. Here's a standard example:

```dart
void main() {
  List<int> xs = [1];
  List<num> numList = xs;
  Iterable<int> intIterable = xs;
  var o = 2 > 1 ? numList : intIterable;
  o.first; // Error `Object` does not have a member named `first`.
}
```

At the time when this part of the algorithm in Dart was designed, other languages including Java had issues with non-termination of several algorithms including 'least upper bound', and hence Dart got a simpler (but safe) algorithm.

Anyway, in the example given here, the inferred type argument of

```dart
  [
    HBaseLeaf(0),
    HBaseChild(
      HBaseLeaf(0),
    ),
    HBaseError("error message"),
  ]
```

is `Object` (which is basically caused by `HBaseError` getting the type argument `HKindInvalid` and the others getting something that involves `HKindValid` by being otherwise unconstrained).

In other words, type inference fails to find a type argument to `HBaseComposite` in the declaration of `invalidExample` because there's too much slack in the bottom-up inference of a type argument for the list literal.

We could handle many examples like this by, e.g., adopting a proposal like https://github.com/dart-lang/language/issues/3282. But we haven't decided on any specific proposal at this time, and it is not an easy problem to resolve, so it might still take some time.

For now, you could specify the actual type argument on the list literal (dealing with the source of the current failure) or the actual type argument of the `HBaseError` instance creation (which is sufficient in order to ensure that the list elements have matching actual type arguments):

```dart
const invalidExample = HBaseComposite(
  [
    HBaseLeaf(0),
    HBaseChild(
      HBaseLeaf(0),
    ),
    HBaseError<HKindValid>("error message"),
  ]
);
```

> ... the "add type annotation" feature has been changed to no longer infer the more precise type.

I do not think this is true. The situation sounds like a completely standard situation for the issue which is "inference failed because the upper bound was too general", and that hasn't changed much since 2011 (OK, lots of cases are being handled much better, but this particular kind of situation has been a problem all the time).

**Comment by modulovalue** [comment]
@eernstg Thank you for the comment!

> I do not think this is true

Just to clarify: I meant to say that when I used to use the "add type annotation" IDE feature, it used to give me a different type from what was being inferred without a type annotation. Now the "add type annotation" feature appears to match what is actually being inferred. There was an inconsistency which made me open this issue, but I can't reproduce that inconsistency anymore and so I've closed this issue.

Everything appears to be working as expected now!
